lightning-invoice/src/payment.rs

```diff
@@ -244,140 +253,166 @@ where
+	/// case a retry is needed.
+	///
+	/// You should ensure that `payment_preimage` is unique and that its `payment_hash` has never
+	/// been paid before. Because [`InvoicePayer`] is stateless no effort is made to do so for you.
+	pub fn pay_pubkey(
+		&self, pubkey: PublicKey, payment_preimage: PaymentPreimage, amount_msats: u64,
+		final_cltv_expiry_delta: u32
+	) -> Result<PaymentId, PaymentError> {
+		let payment_hash = PaymentHash(Sha256::hash(&payment_preimage.0).into_inner());
+		match self.payment_cache.lock().unwrap().entry(payment_hash) {
+			hash_map::Entry::Occupied(_) => return Err(PaymentError::Invoice("payment pending")),
+			hash_map::Entry::Vacant(entry) => entry.insert(0),
+		};
+
+		let params = RouteParameters {
+			payee: Payee::for_keysend(pubkey),
+			final_value_msat: amount_msats,
+			final_cltv_expiry_delta,
+		};
+
+		let send_payment = |route: &Route| {
+			self.payer.send_spontaneous_payment(route, payment_preimage)
+		};
+		self.pay_internal(&params, payment_hash, send_payment)
+			.map_err(|e| { self.payment_cache.lock().unwrap().remove(&payment_hash); e })
+	}
+
+	fn pay_internal<F: FnOnce(&Route) -> Result<PaymentId, PaymentSendFailure> + Copy>(
+		&self, params: &RouteParameters, payment_hash: PaymentHash, send_payment: F,
+	) -> Result<PaymentId, PaymentError> {
+		if has_expired(params) {
+			log_trace!(self.logger, "Invoice expired prior to send for payment {}", log_bytes!(payment_hash.0));
 			return Err(PaymentError::Invoice("Invoice expired prior to send"));
 		}
-		let retry_data_payment_id = loop {
-			let mut payment_cache = self.payment_cache.lock().unwrap();
-			match payment_cache.entry(payment_hash) {
-				hash_map::Entry::Vacant(entry) => {
-					let payer = self.payer.node_id();
-					let mut payee = Payee::from_node_id(invoice.recover_payee_pub_key())
-						.with_expiry_time(expiry_time_from_unix_epoch(&invoice).as_secs())
-						.with_route_hints(invoice.route_hints());
-					if let Some(features) = invoice.features() {
-						payee = payee.with_features(features.clone());
+
+		let payer = self.payer.node_id();
+		let first_hops = self.payer.first_hops();
+		let route = self.router.find_route(
+			&payer, params, &payment_hash, Some(&first_hops.iter().collect::<Vec<_>>()),
+			&self.scorer.lock()
+		).map_err(|e| PaymentError::Routing(e))?;
+
+		match send_payment(&route) {
+			Ok(payment_id) => Ok(payment_id),
+			Err(e) => match e {
+				PaymentSendFailure::ParameterError(_) => Err(e),
+				PaymentSendFailure::PathParameterError(_) => Err(e),
+				PaymentSendFailure::AllFailedRetrySafe(_) => {
+					let mut payment_cache = self.payment_cache.lock().unwrap();
+					let retry_count = payment_cache.get_mut(&payment_hash).unwrap();
+					if *retry_count >= self.retry_attempts.0 {
+						Err(e)
+					} else {
+						*retry_count += 1;
+						std::mem::drop(payment_cache);
+						Ok(self.pay_internal(params, payment_hash, send_payment)?)
```

Comment:
I'd think this would call `self.retry_payment` (though maybe it doesn't make a difference?), maybe a comment as to the reasoning?

Reply:
We can't call `retry_payment` because we would need to pass it a `RouteParameter`, which is only given for partial failures.